### ISSUE TITLE
Turn on C++17 in order to use if constexpr

### ISSUE THREAD

**Issue by fcooper8472** [Chaste/Chaste#107]
Looking to move quite quickly on this.

## The problem

One of the things breaking the build at the moment is newer compilers quite correctly complaining about accessing elements of c_vector that are out of range. This is, for instance, caused when we do:

```cpp
assert(SPACE_DIM == 2);

// code here

c_vector<double, SPACE_DIM> vector_a_to_b = GetVectorFromAtoB(vertexA, vertexB);

if (vector_a_to_b[1] < DBL_EPSILON)
{
    // do something
}
```

This code is wrong and always has been - the function is getting compiled for `SPACE_DIM` being 1, 2 and 3, and in the case where it's 1, then `vector_a_to_b[1] ` is genuinely out of range. The compiler is right to emit a warning (which we treat as an error).

This happens all over the code. We have done various things to fix it in the past, but there are places (such as in `VertexMesh`) where no fix is really satisfactory.

## Proposed solution

[`if constexpr`](https://en.cppreference.com/w/cpp/language/if) is a feature in C++17. The idea is to perform conditional compilation based on a condition that is known at compile time. In other words, it's a way to tell the compiler "if this condition is true, compile this block of code; otherwise, ignore it."

This would allow us to write code like:

```cpp
template <unsigned ELEMENT_DIM, unsigned SPACE_DIM>
c_vector<double, SPACE_DIM> VertexMesh<ELEMENT_DIM, SPACE_DIM>::GetShortAxisOfElement(unsigned index)
{
    if constexpr (SPACE_DIM == 2)
    {
        // Code that only makes sense in 2d
    }
    else
    {
        NEVER_REACHED;
    }
}
```

## Benefits

1. Short of refactoring all of our classes with, e.g., partial class template specialization, this is the best solution I believe exists to address this problem.
2. Code that is currently being compiled up to 6 times (`<1, 1>, <1, 2>, <1, 3>, <2, 2>, <2, 3>, <3, 3>`) will actually only be compiled when needed, i.e. a 3x compilation speedup for `SPACE_DIM == 2`.

## Short turnaround for decision

I would like to do this ASAP in an effort to get @AlexFletcher his green ticks on the build before the workshop, and I would much rather actually fix the underlying problem, rather than fudge it.

**Comment by fcooper8472** [comment]
See #108 for an example of what this looks like in practice.

I'm expecting this to have resolved the build errors present here:
https://github.com/Chaste/Chaste/actions/runs/5253369870/jobs/9490611504

**Comment by fcooper8472** [comment]
Closed by #108
